### DISCUSSION TITLE
Podman ?

### DISCUSSION THREAD

**Discussion by carlocorradini** [k3s-io/k3s#4277] [Q&A] [answered]
It's possible to use [Podman](https://podman.io) as container runtime?

**Comment by brandond** [comment, marked as answer]
As far as I know podman doesn't provide a CRI socket for the Kubelet to connect to, since it's just a daemonless replacement for the CLI portion of docker. I believe you would need to use something else like cri-o.

**Reply by carlocorradini** [reply]
Thanks!

**Comment by mtalexan** [comment]
Some background for future searchers of this same topic:

CRI is the interface needed by k3s/k8s.  
`docker` has 2 parts to it, both under the same name: a CLI client and a daemon.  
`containerd` is a daemon only, though there's a related simplistic CLI tool, `ctr`, that can do some very simple things with it.   
`podman` has a primarly daemonless rootless CLI client, but can optionally also run a daemon and interact with that daemon using the normally daemonless rootless CLI client (with the `--remote` option).  The optional daemon is intended to be used as a replacement for the docker daemon and attempts to be a fully compatible daemon for even the docker client.  
CRI-O (`crio`) is a tool developed by and sharing design/code with `podman` that is primarily a daemon, though it has `crictl` for some interaction.  

The docker daemon supports 2 different endpoint interface sets from the same socket: CRI and the CLI client interface. 
The containerd daemon only supports CRI.   
The optional podman daemon supports only the docker CLI client interface with some minor additions that only the podman CLI knows how to use if you use `--remote`.  
`crio` only supports CRI (from what I can tell).  


CRI-O and podman use the same image cache storage technology, and it's possible to use `podman --root=...` to point at the CRI-O image storage location to manage the cache using the podman CLI.  Supposedly you can also use the podman CLI to talk to the CRI-O daemon as well, though I've yet to find documentation on how to do this (maybe `--remote`?).

---


K8s and K3s technically only need a daemon supporting the CRI interface.  Historically this was `dockerd` or `containerd`, but `dockerd` support has been deprecated/removed from K8s.  At some point the default backend for `K8s` was changed to CRI-O.  
K3s also deprecated `dockerd` when K8s did, but as of now doesn't officially support CRI-O as a backend.  One of the benefits of k3s is that it packages what you need with it, so it has `containerd` packaged with it still and used as the default.  

It is technically possible, though not directly supported, to use CRI-O as the K3s backend if you setup CRI-O yourself and then point the K3s server at it with the `--container-runtime-endpoint` option, but it's not a Rancher supported configuration.
